src/apps/external_executor/external_executor.cpp

```diff
@@ -207,47 +227,90 @@ namespace externalexecutor
+          LOG_INFO_FMT("Activated executor {}", executor_id);
+
+          // Update dispatch map with this executor
+          const auto& uris = supported_uris[executor_id];
+          for (const auto& uri : uris)
+          {
+            supported_uris_for_active_executors[uri].insert(executor_id);
+          }
+
+          return ccf::grpc::make_pending();
+        }
+      };
+      make_endpoint(
+        "/externalexecutor.protobuf.KV/Activate",
+        HTTP_POST,
+        ccf::grpc_command_unary_stream_adapter<
+          google::protobuf::Empty,
+          externalexecutor::protobuf::Work>(activate),
+        executor_only)
+        .set_forwarding_required(ccf::endpoints::ForwardingRequired::Never)
+        .install();
+
+      auto deactivate = [this](
+                          ccf::endpoints::CommandEndpointContext& ctx,
+                          google::protobuf::Empty&& payload)
+        -> ccf::grpc::GrpcAdapterResponse<google::protobuf::Empty> {
         const auto executor_id = get_caller_executor_id(ctx);
-        const auto it = active_requests.find(executor_id);
-        if (it != active_requests.end())
+        const auto it = active_executors.find(executor_id);
+        if (it == active_executors.end())
         {
           return ccf::grpc::make_error(
             GRPC_STATUS_FAILED_PRECONDITION,
-            "Already managing an active transaction");
+            fmt::format("Executor {} was not active", executor_id));
         }
 
-        externalexecutor::protobuf::OptionalRequestDescription
-          optional_request_description;
+        // Signal to this executor that its work has finished
+        externalexecutor::protobuf::Work work;
+        work.mutable_work_done();
+        it->second.work_stream->stream_msg(work);
 
-        auto& executor_queue = pending_executor_requests[executor_id];
-        if (!executor_queue.empty())
+        active_executors.erase(it);
+        LOG_INFO_FMT("Deactivated executor {}", executor_id);
+
+        for (auto& [uri, executors_list] : supported_uris_for_active_executors)
         {
-          auto* request_description =
-            optional_request_description.mutable_optional();
-          auto pending_request = executor_queue.front();
-          executor_queue.pop();
-          LOG_TRACE_FMT(
-            "Processing executor id:{}, uri: {}, method: {}",
-            executor_id.value(),
-            pending_request->request_description.uri(),
-            pending_request->request_description.method());
-          *request_description = pending_request->request_description;
-          active_requests.emplace_hint(it, executor_id, pending_request);
+          executors_list.erase(executor_id);
         }
 
-        return ccf::grpc::make_success(optional_request_description);
+        return ccf::grpc::make_success();
       };
       make_endpoint(
-        "/externalexecutor.protobuf.KV/StartTx",
+        "/externalexecutor.protobuf.KV/Deactivate",
```

Comment:
I think this `EndTx` endpoint logic can also be combined with this?

Reply:
I don't think so. `Deactivate` is once-per-executor (paired with a single call to `Activate`), whereas `EndTx` should be called multiple times by each executor (once-per-request, paired with each message from `Activate`).

If we supported async client streaming, then this would probably be an additional message in the client->server direction (as it is for the etcd watch API), but since that's not supported we implement it as a standalone unary call.

I may also remove this RPC entirely, and say that client-initiated closures just terminate their `Activate` RPC session. That changes the expected client flow if clients want to be able to do this (ie - our e2e tests), and makes it harder to distinguish deliberate/polite executor closures from crashing/failure executor closures, but that might not be a problem.

Reply:
> Deactivate is once-per-executor (paired with a single call to Activate), whereas EndTx should be called multiple times by each executor (once-per-request

Ah yes, my bad!

> makes it harder to distinguish deliberate/polite executor closures from crashing/failure executor closures, but that might not be a problem.

Yeah, I believe this can be handled with an appropriate error response